### ISSUE TITLE
WebSocket handshake not working when using Sec-WebSocket-Protocol header

### ISSUE THREAD

**Issue by rschlegel-cc** [http4k/http4k#454]
Hi,

recently I started migrating a small service from ktor to http4k and experienced some handshake issues when using the http4k-client-websocket (3.251.0) module, despite sending the exact same request to start the session as in the previous implementation (verified in Wireshark). 

The code used to create the non-blocking client:
```kotlin
WebsocketClient.nonBlocking(Uri.of("ws://192.168.1.24:8214"), listOf("Sec-WebSocket-Protocol" to "Lux_WS")) {
    // not relevant, after handshake
}
```

This led to the error message `1002 draft Draft_6455 extension: DefaultExtension max frame size: 2147483647 refuses handshake` in the onError-handler. After some digging around I found out that the value passed in the Sec-WebSocket-Protocol needs to be included as a protocol when creating the Draft_6455 instance. I tried replacing `Draft_6455()` with `Draft_6455(emptyList<IExtension>(), headers.filter { "Sec-WebSocket-Protocol".equals(it.first, true) }.map {Protocol(it.second)})` in WebsocketClient.kt:83 and while it is surely not the most elegant solution it does indeed allow the handshake to complete.

Is there a better way of achieving this that I'm missing or is this a rather obscure edge case?

**Comment by daviddenton** [comment]
Suspect this may be an edgecase, but am not an expert on this. Since this Draft concept is part of the underlying client API that we're using, the only thing I can see which looks suspicious is the change from the current no-args constructor which currently (effectively) does this:

```	
public Draft_6455(  ) {
		this( Collections.<IExtension>emptyList(), Collections.<IProtocol>singletonList( new Protocol( "" ) ));
	}
```

I've slightly refactored out the websocket classes to make this a bit simpler in master in case we want to make a change here, but maybe ask in the https://github.com/TooTallNate/Java-WebSocket project and see what they say there? 

If we do want to make a change, then I'd prefer not to expose the underlying API outside if at all possible - but this might not be workable so we'll have to see. As it stands, with the change I've just made, you'd only to just implement your own websocket client (3-5 lines) to use the custom Draft with the http4k classes.

**Comment by rschlegel-cc** [comment]
Thanks for the quick response. You're right, I messed up the default case with the workaround I posted. I tried the proposed change and as far as I can tell it fixes my problem in a way that is consistent with the Java-WebSocket documentation (https://github.com/TooTallNate/Java-WebSocket/wiki/Specifing-a-Sec-WebSocket-Protocol) while keeping the API simple for more standard cases.

**Comment by daviddenton** [comment]
If you fancy PRing the change from the latest master then we'll get it merged in. :)

**Comment by rschlegel-cc** [comment]
Sure, I'd love to, where does it need to go to? Normally I would expect to create a PR to master, but the commit already in there. I hope I'm not missing something really obvious here.

**Comment by daviddenton** [comment]
I've exposed the Draft now so that people don't have to implement the entire client again if they want to use a custom protocol. It will be released in 3.253.0 whenever that lands. :) 

Closing for now. :)
